src/Constraint/ConstraintInterface.php

```diff
@@ -11,15 +11,31 @@
 
 namespace Composer\Semver\Constraint;
 
+/**
+ * DO NOT IMPLEMENT this interface. It is only meant for usage as a type hint
+ * where appropriate but we do not support third parties implementing this
+ * interface themselves, and will do BC breaks to the interface as we see fit.
+ */
```

Comment:
```suggestion
 * @internal
 */
```

Reply:
Yeah my concern with `@internal` was that IDEs would complain if the interface is used as type hint outside the package. If that's not the case I happily add that though. I just have no idea how PHPStorm behaves exactly.

Reply:
I think the message is misleading. You should be type hinting towards this interface and thus `and will do BC breaks to the interface as we see fit` is incorrect. What you shouldn't do is invent your own constraint that implements this interface. I'd write something along the lines of
```
 * DO NOT IMPLEMENT this interface. It is only meant for usage as a type hint
 * in libraries relying on composer/semver but creating your own constraint class
 * that implements this interface is not a supported use case and will cause the
 * composer/semver components to return unexpected results.
```

Reply:
I know we're technically doing `BC breaks to the interface as we see fit` by adding a method like you did. But we cannot just remove one, that's why I wouldn't write that :)

Reply:
@Toflar fair point thanks